src/Features/Core/Diagnostics/BaseDiagnosticIncrementalAnalyzer.cs

```diff
@@ -19,24 +19,145 @@ protected BaseDiagnosticIncrementalAnalyzer(Workspace workspace)
+        /// <param name="document"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
         public abstract Task DocumentOpenAsync(Document document, CancellationToken cancellationToken);
+        /// <summary>
+        /// Flush cached diagnostics produced by a prior analysis of a document.
+        /// </summary>
+        /// <param name="document"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
         public abstract Task DocumentResetAsync(Document document, CancellationToken cancellationToken);
+        /// <summary>
+        /// This method appears not to need to have an effect.
+        /// </summary>
+        /// <param name="solution"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
         public abstract Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancellationToken);
+        /// <summary>
+        /// Flush diagnostics produced by a prior analysis of a document,
+        /// and suppress future analysis of the document.
+        /// Calls <see cref="DiagnosticAnalyzerService.RaiseDiagnosticsUpdated(object, DiagnosticsUpdatedArgs)"/> with an empty collection.
+        /// </summary>
+        /// <param name="documentId"></param>
         public abstract void RemoveDocument(DocumentId documentId);
+        /// <summary>
+        /// Flush diagnostics produced by a prior analysis of a project,
+        /// and suppress future analysis of the project.
+        /// Calls <see cref="DiagnosticAnalyzerService.RaiseDiagnosticsUpdated(object, DiagnosticsUpdatedArgs)"/> with an empty collection.
+        /// </summary>
+        /// <param name="projectId"></param>
         public abstract void RemoveProject(ProjectId projectId);
         #endregion
 
         #region delegating methods from diagnostic analyzer service to each implementation of the engine
+        /// <summary>
+        /// Get previously-computed (and potentially stale) diagnostics associated with a particular combination of
+        /// analysis classification (syntax/semantics/project), document/project, and analyzer.
+        /// </summary>
+        /// <param name="solution"></param>
+        /// <param name="id">Matched against values supplied in a <see cref="DiagnosticsUpdatedArgs"/> to <see cref="DiagnosticAnalyzerService.RaiseDiagnosticsUpdated(object, DiagnosticsUpdatedArgs)"/>.</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
         public abstract Task<ImmutableArray<DiagnosticData>> GetSpecificCachedDiagnosticsAsync(Solution solution, object id, CancellationToken cancellationToken);
+        /// <summary>
+        /// Get previously-computed (and potentially stale) diagnostics associated with a particular document, project, or solution.
+        /// </summary>
+        /// <param name="solution">Ignored if projectId or documentId is non null.</param>
```

Comment:
OK, so "ignored" is obviously incorrect. In digging through the logic of the current implementation, it appears as if there is a rule in place of:
-- If a document is supplied, return diagnostics for that document.
-- Otherwise, if a project is supplied, return diagnostics for that project.
-- Otherwise, return diagnostics for the solution.
Is this correct? Assuming so, I'll amend the doc comments to be closer to correct.